### ISSUE TITLE
Allow to add notes to Sites Nodes

### ISSUE THREAD

**Issue by MinhPham123456789** [zaproxy/zaproxy#8240]
### Is your feature request related to a problem? Please describe.

I believe this feature can highly support user to **attach relevant notes directly** to a directory path or a particular domain **in the sites tree** and retrieve the notes more efficiently than having the notes attached to separate requests in the History. In addition, **having a note attached to a node in the sites tree** allows users to treat Zaproxy as a second place to store the pentesting notes, because the notes are directly attached to the relevant targeted directory path or URL or domain.

### Describe the solution you'd like

The user can:
1) The user can right-click on a node in the sites tree
2) The pop-up options window shows an option to create a note "Add note"
3) The user clicks on the "Add note" option
4) The note is displayed as a tab in the right panel next to the Request and Response tab
5) The user can create, edit, or empty the Note
6) The user can close the Note tab normally just like closing the Requester tab

### Describe alternatives you've considered

Currently besides the note in individual requests in the History, there is no other alternative. I think the note attached to each request in the History is not a good approach to support the users using Zaproxy in an efficient way

### Would you like to help fix this issue?

- [ ] Yes

**Comment by amitpanwar789** [comment]
This looks interesting.

**Comment by psiinon** [comment]
@amitpanwar789 if you would like to work on it then just say.
If you dont want to work on it but would like to see it implemented then give the first comment a 👍 - we use those to determine which issues are most popular..

**Comment by amitpanwar789** [comment]
Yes, I will like to work on this.

**Comment by psiinon** [comment]
@amitpanwar789 I've assigned it to you 😁 
You are right in that it is non trivial to implement, however it shouldnt be too bad.
I would start by showing exiting (History) notes in the Sites Tree as well - there are methods for adding icons to site notes. Then you can add a right click option for viewing them, and then for adding them.

**Comment by thc202** [comment]
That seems to suggest the notes will be added to history entries rather than the sites nodes as wanted by this issue.

**Comment by amitpanwar789** [comment]
After reviewing the existing notes feature in the request history, I believe we can implement a similar functionality for SiteNodes within the Site Tree. This may involve creating an ExtensionSiteNode class that incorporates components like NotesAddDialog and PopupMenuNote. However, a key decision point remains: should we completely rewrite these components specifically for SiteNodes, or can we modify the existing ones to be more generic? For example, we could modify NotesAddDialog to accept SiteNode instead of HistoryReference for SiteNode, and similarly, ExtensinoSiteNode could replace ExitensionHIstory in PopupMenuNote.

**Comment by psiinon** [comment]
So .. what I was suggesting was a quick hack .. but I think its still valid 😉 
The notes would still be stored in the history refs, but as far as the user is concerned they are still also shown in the Sites tree.
The one time where they need to know they are still accociated with hrefs is if they delete the relevant href, then the Site Node Note also disappears. But there are ways around that, ir warning the user that an href has a note attached when they try to delete it.

**Comment by amitpanwar789** [comment]
but sometime a siteNode has multiple href, assume all of them have there respective notes and if the user access them from siteNode then which note the user should see how will we decide this or we can somehow show all the notes by showing the respective href with a note.

**Comment by psiinon** [comment]
Good point, that does make it harder.
I guess we could have one icon per note, but that could get very messy.
The alternative is to forget about my suggestion hack and to implement the full solution from the start...

**Comment by amitpanwar789** [comment]
> Good point, that does make it harder. I guess we could have one icon per note, but that could get very messy. The alternative is to forget about my suggestion hack and to implement the full solution from the start...

i have a idea, maybe we can make a whole new Note box, which work like, whenever we open it from SiteNode it will show all the history reference associate with that SiteNode and user first have to select the history reference then he can add/modify/delete note for that respective history reference , this way we don't have to implement any additional storage as told by kingthorin 'Sitenodes are created by ZAP and not really stored' .

**Comment by amitpanwar789** [comment]
@psiinonIi just want to clarify if my idea is right or not or want suggestions before implementing it, so i was thinking like adding a NotesAddDialog, PopupMenuNote Class in SiteMap.java so that i can pass a SiteNode to NotesAddDialog and NotesAddDialog box will have two section first one will contain all the href associated to that SiteNode( like href from its children SiteNode and similar href with same address) and second section is text area for note.

**Comment by amitpanwar789** [comment]
How can i put add Notes options in PopupMenu options of siteMap node, by studing the other options as all the options are made by extending extension adaptor , do i need to do the same thing ? and if yes can you suggest me some extension which are small and ease to understand.

**Comment by thc202** [comment]
It's not necessary add a new extension, IMO, it can be added to `ExtensionHistory`.

**Comment by amitpanwar789** [comment]
Yes i tried to add a dummy options to popupMenu of ExtenstionHistory but it will only show if we select the HistoryReference through history panel not from SiteTree , if a right click on siteNode that dummy options is not in the list of options ?

**Comment by thc202** [comment]
Which class are you using for the context menu? e.g. `PopupMenuExportURLs` would show in the sites tree (see `isEnableForComponent`).

**Comment by amitpanwar789** [comment]
i am just adding this line in ExtensionHistory.java file (PopupMenuSiteNode is a dummy options extend from PopupMenuItemHistoryReferenceContainer)

`extensionHook.getHookMenu().addPopupMenuItem( new PopupMenuSiteNode(this));`

**Comment by amitpanwar789** [comment]
One thing I want to ask is, as I already told you, I'm going to make a list of all the History References related to a SiteNode in the left panel of the NotesDialog box. However, I'm unsure what text to display for each list item so that users can uniquely identify it. Initially, I thought URL + TimeStamp would work, but URLs can be quite long, potentially making the UI inconsistent. What would you recommend using instead?

**Comment by thc202** [comment]
The ID of the history reference.

**Comment by thc202** [comment]
Override `isEnableForInvoker` to only return true for `SITES_PANEL` (note that if you are extending that class you don't need to override `isEnableForComponent`).

**Comment by amitpanwar789** [comment]
I'm writing to follow up on my PR and  I'm aware you're likely busy, see if you had a chance to take a look at it . I'm waiting for your feedback
